File: disc-math/3Semestr/lab1/H.java
import java.util.*;

class Graph {
    private List<Set<Integer>> vertexes;
    private int size;
    private int countEdges;

    public Graph(List<Set<Integer>> vertexes) {
        this.vertexes = vertexes;
        this.size = vertexes.size();

        for (Set<Integer> edges : vertexes) {
            countEdges += edges.size();
        }
    }

    public Graph join(int u, int v) {
        int min = Math.min(u, v);
        int max = Math.max(u, v);

        List<Set<Integer>> newVertexes = new ArrayList<>();

        for (int i = 0; i < size; ++i) {
            if (i == min) {
                HashSet<Integer> filtred = new HashSet<>();
                for (int el : vertexes.get(min)) {
                    if (el != max) {
                        filtred.add(el);
                    }
                }
                newVertexes.add(filtred);
            } else if (i == max) {
                HashSet<Integer> filtred = new HashSet<>();
                for (int el : vertexes.get(max)) {
                    if (el != min) {
                        filtred.add(el);
                    }
                }
                newVertexes.get(min).addAll(filtred);
            } else {
                HashSet<Integer> filtred = new HashSet<>();
                for (int el : vertexes.get(i)) {
                    if (el == max) {
                        filtred.add(min);
                    } else {
                        filtred.add(el);
                    }
                }
                newVertexes.add(filtred);
            }
        }

        return new Graph(newVertexes);
    }

    public Graph split(int u, int v) {
        List<Set<Integer>> newVertexes = new ArrayList<>();
        for (int i = 0; i < size; ++i) {
            if (i == u || i == v) {
                HashSet<Integer> filtred = new HashSet<>();
                for (int el : vertexes.get(i)) {
                    if (el != u && el != v) {
                        filtred.add(el);
                    }
                }
                newVertexes.add(filtred);
            } else {
                newVertexes.add(vertexes.get(i));
            }
        }

        return new Graph(newVertexes);
    }

    public List<Set<Integer>> getVertexes() {
        return vertexes;
    }

    public int getSize() {
        return size;
    }

    public int getCountEdges() {
        return countEdges;
    }
}

class Polinom {
    private List<Monom> items;
    private int size;

    public Polinom(List<Monom> items) {
        this.items = items;
        this.size = items.size();
    }

    public Polinom(Monom monom) {
        List<Monom> res = new ArrayList<>();
        for (int i = 0; i < monom.getDegree(); ++i) {
            res.add(new Monom(i, 0));
        }
        res.add(monom);

        this.items = res;
        this.size = res.size();
    }

    public Polinom() {
        this(new ArrayList<>());
    }

    public Polinom add(Monom monom) {
        return add(new Polinom(monom));
    }

    public Polinom add(Polinom other) {
        int newSize = Math.max(size, other.getSize());
        List<Monom> newItems = new ArrayList<>();

        for (int i = 0; i < newSize; ++i) {
            newItems.add(new Monom(i, 0));
        }

        for (int i = 0; i < newSize; ++i) {
            if (i < size) {
                newItems.set(i, newItems.get(i).add(items.get(i)));
            }
            if (i < other.getSize()) {
                newItems.set(i, newItems.get(i).add(other.getItems().get(i)));
            }
        }

        return new Polinom(newItems);
    }

    public void multiplyOnConst(int c) {
        for (Monom monom : items) {
            monom.setC(monom.getC() * c);
        }
    }

    public int eval(int value) {
        int res = 0;

        for (Monom monom : items) {
            res += monom.eval(value);
        }

        return res;
    }

    public List<Monom> getItems() {
        return items;
    }

    public void setItems(List<Monom> items) {
        this.items = items;
    }

    public int getSize() {
        return size;
    }

    @Override
    public String toString() {
        StringBuilder res = new StringBuilder();
        for (int i = items.size() - 1; i >= 0; --i) {
            res.append(items.get(i).toString());
        }

        return res.toString();
    }
}

class Monom {
    private int degree;
    private int c;

    public Monom(int degree) {
        this(degree, 1);
    }

    public Monom(int degree, int c) {
        this.degree = degree;
        this.c = c;
    }

    public Monom add(Monom other) {
        if (degree != other.getDegree()) {
            return null;
        }

        return new Monom(degree, c + other.c);
    }

    public int eval(int value) {
        return c * ((int) Math.pow(value, degree));
    }
    
    public int getDegree() {
        return degree;
    }

    public void setDegree(int degree) {
        this.degree = degree;
    }

    public int getC() {
        return c;
    }

    public void setC(int c) {
        this.c = c;
    }

    @Override
    public String toString() {
        String constStr = c >= 0 ? "+" + c : Integer.toString(c);
        return constStr + "*t^" + degree;
    }
}

public class H {
    public static void main(String[] args) {
        new H().solve();
    }

    void solve() {
        Scanner in = new Scanner(System.in);

        int n = in.nextInt();
        int m = in.nextInt();

        List<Set<Integer>> vertexes = new ArrayList<>();
        for (int i = 0; i < n; ++i) {
            vertexes.add(new HashSet<>());
        }

        for (int i = 0; i < m; ++i) {
            int from = in.nextInt() - 1;
            int to = in.nextInt() - 1;

            vertexes.get(from).add(to);
            vertexes.get(to).add(from);
        }

        in.close();

        Graph graph = new Graph(vertexes);
        Polinom res = recur(graph);

        System.out.println(res.getItems().size() - 1);

        List<Monom> resPolinom = res.getItems();
        for (int i = resPolinom.size() - 1; i >= 0; --i) {
            System.out.print(resPolinom.get(i).getC());
            System.out.print(" ");
        }
    }

    Polinom recur(Graph graph) {
        if (graph.getCountEdges() == 0) {
            return new Polinom(new Monom(graph.getSize()));
        } else {
            int u = 0, v = 0;
            for (int i = graph.getSize() - 1; i >= 0; --i) {
                if (!graph.getVertexes().get(i).isEmpty()) {
                    u = i;
                    v = (int) graph.getVertexes().get(i).toArray()[graph.getVertexes().get(i).size() - 1];

                    break;
                }
            }

            Polinom resSplittedGraph = recur(graph.split(u, v));
            Polinom resJoinedGraph = recur(graph.join(u, v));
            resJoinedGraph.multiplyOnConst(-1);

            return resSplittedGraph.add(resJoinedGraph);
        }
    }
}

File: disc-math/3Semestr/lab1/H.cpp
#include <iostream>
#include <vector>
#include <set>

using namespace std;

using graph_type = vector<set<int>>;

struct Graph {
    graph_type vertexes;
    int size;
    int count_edges;

    Graph(const graph_type &vertexes) {
        this->vertexes = vertexes;
        this->size = vertexes.size();

        this->count_edges = 0;
        for (auto edges: vertexes) {
            this->count_edges += edges.size();
        }
        this->count_edges /= 2;
    }

    Graph join(int u, int v) {
        int min_v = min(u, v);
        int max_v = max(u, v);

        vector<bool> used(size, false);
        graph_type new_vertexes(size - 1);

        for (auto el: vertexes[max_v]) {
            int index = el > max_v ? el - 1 : el;

            new_vertexes[index] = vertexes[el];
            new_vertexes[index].erase(max_v);
            if (el != min_v) {
                new_vertexes[index].insert(min_v);
                used[el] = true;
            }
        }

        used[min_v] = true;
        used[max_v] = true;

        for (int i = 0; i < size; ++i) {
            if (!used[i]) {
                int index = i > max_v ? i - 1 : i;
                new_vertexes[index] = vertexes[i];
            }
        }

        return Graph(new_vertexes);
    }

    Graph split(int u, int v) {
        graph_type new_vertexes = this->vertexes;
        new_vertexes[u].erase(v);
        new_vertexes[v].erase(u);

        return Graph(new_vertexes);
    }
};

struct Polinom {
    vector<int> items;
    int degree;

    Polinom(const vector<int> &items) {
        this->items = items;
        this->degree = items.size() - 1;
    }

    Polinom add(const Polinom &other) {
        int new_size = max(degree, other.degree) + 1;
        vector<int> new_items(new_size, 0);

        for (int i = 0; i < new_size; ++i) {
            if (i <= degree) {
                new_items[i] += items[i];
            }
            if (i <= other.degree) {
                new_items[i] += other.items[i];
            }
        }

        return Polinom(new_items);
    }

    void multiply_on_const(int c) {
        for (int i = 0; i < items.size(); ++i) {
            items[i] *= c;
        }
    }
};

Polinom recur(Graph graph) {
    if (graph.count_edges == 0) {
        vector<int> args(graph.size, 0);
        args.push_back(1);

        return Polinom(args);
    } else {
        int u = 0;
        int v = 0;

        for (int i = graph.size - 1; i >= 0; --i) {
            if (!graph.vertexes[i].empty()) {
                u = i;
                v = *(--graph.vertexes[i].end());

                break;
            }
        }

        Polinom res_spl_graph = recur(graph.split(u, v));
        Polinom res_join_graph = recur(graph.join(u, v));

        res_join_graph.multiply_on_const(-1);

        return res_spl_graph.add(res_join_graph);
    }
}

void solve() {
    int n, m;
    cin >> n >> m;
    graph_type vertexes(n);

    for (int i = 0; i < m; ++i) {
        int from, to;
        cin >> from >> to;
        --from;
        --to;

        vertexes[from].insert(to);
        vertexes[to].insert(from);
    }

    Graph graph(vertexes);
    Polinom res = recur(graph);

    cout << res.degree << "\n";
    for (int i = res.items.size() - 1; i >= 0; --i) {
        cout << res.items[i] << " ";
    }
}

int main() {
    solve();
}
